lib/puppet/configurer.rb

```diff
@@ -102,32 +102,44 @@ def retrieve_catalog(query_options)
   end
 
   # Convert a plain resource catalog into our full host catalog.
-  def convert_catalog(result, duration)
-    catalog = result.to_ral
-    catalog.finalize
-    catalog.retrieval_duration = duration
-    catalog.write_class_file
-    catalog.write_resource_file
+  def convert_catalog(result, duration, options = {})
+    catalog = nil
+
+    catalog_converstion_time = thinmark do
+      catalog = result.to_ral
+      catalog.finalize
+      catalog.retrieval_duration = duration
+      catalog.write_class_file
+      catalog.write_resource_file
+    end
+    options[:report].add_times(:convert_catalog, catalog_converstion_time) if options[:report]
+
     catalog
   end
 
   def get_facts(options)
     if options[:pluginsync]
-      remote_environment_for_plugins = Puppet::Node::Environment.remote(@environment)
-      download_plugins(remote_environment_for_plugins)
+      plugin_sync_time = thinmark do
+        remote_environment_for_plugins = Puppet::Node::Environment.remote(@environment)
+        download_plugins(remote_environment_for_plugins)
 
-      Puppet::GettextConfig.reset_text_domain('agent')
-      Puppet::ModuleTranslations.load_from_vardir(Puppet[:vardir])
+        Puppet::GettextConfig.reset_text_domain('agent')
+        Puppet::ModuleTranslations.load_from_vardir(Puppet[:vardir])
+      end
+      options[:report].add_times(:plugin_sync, plugin_sync_time) if options[:report]
```

Comment:
Looks good. One nit is that the agent may make multiple passes if the environment doesn't match. So maybe we need to accumulate `plugin_sync`, `fact_generation`, and `node_retrieval` times? e.g `options[:report].add_times(:plugin_sync, pluginsync_time, true)` and add an optional arg to `Report#add_times` to either overwrite or accumulate the metric?

Reply:
I was actually just thinking about trying to figure out what would happen in that case. Yes, that sounds like a good solution here. I'll look into adding that in.

Reply:
This is done, but I modified the `add_times` method to automatically accrue metrics unless otherwise specified. I think we don't want to accidentally overwrite time data just in case we guess wrong  about which methods are called multiple times.